logstash-core/src/main/java/org/logstash/util/CloudSettingId.java

```diff
@@ -0,0 +1,172 @@
+            if (hostParts.length > 1) {
+                port = hostParts[1];
+            }
+            return new HostAndPort(host, port);
+        }
+    }
+
+    public static final String DOT_SEPARATOR = ".";
+    public static final String CLOUD_PORT = "443";
+
+    private String original;
+    private String decoded;
+    private String label;
+    private String elasticsearchScheme;
+    private String elasticsearchHost;
+    private String elasticsearchPort;
+    private String kibanaScheme;
+    private String kibanaHost;
+    private String kibanaPort;
+    private String[] otherIdentifiers = new String[0];
+
+    public CloudSettingId(String value) {
+        if (value == null) {
+            return;
+        }
+        original = value;
+        final String[] parts = original.split(":");
+        label = parts[0];
+        String encoded = null;
+        if (parts.length > 1) {
+            encoded = parts[1];
+        }
+        if (encoded == null || encoded.isEmpty()) {
+            try {
+                decoded = new String(Base64.getUrlDecoder().decode(label), StandardCharsets.UTF_8);
+            } catch (IllegalArgumentException iaex) {
+                decoded = "";
+            }
+            label = "";
+        } else {
+            try {
+                decoded = new String(Base64.getUrlDecoder().decode(encoded), StandardCharsets.UTF_8);
+            } catch (IllegalArgumentException iaex) {
+                decoded = "";
+            }
+        }
+        long separatorCount = decoded.chars().filter(c -> c == '$').count();
+        if (separatorCount < 2) {
+            throw new IllegalArgumentException("Cloud Id, after decoding, is invalid. Format: '<segment1>$<segment2>$<segment3>'. Received: \"" + decoded + "\".");
+        }
+        final String[] segments = decoded.split("\\$");
+        if (Arrays.stream(segments).anyMatch(String::isEmpty)) {
+            throw new IllegalArgumentException("Cloud Id, after decoding, is invalid. Format: '<segment1>$<segment2>$<segment3>'. Received: \"" + decoded + "\".");
+        }
+        String cloudBase = segments[0];
+        String cloudHost = DOT_SEPARATOR + cloudBase;
+        final String[] hostParts = cloudHost.split(":");
+        final HostAndPort cloud = new HostAndPort(hostParts[0], hostParts.length > 1 ? hostParts[1] : CLOUD_PORT);
+
+        final HostAndPort elasticsearch = HostAndPort.parseHostAndPort(segments[1], "Cloud Id, after decoding, elasticsearch segment is 'undefined', literally.");
+        elasticsearchPort = elasticsearch.portOrDefault(cloud.port);
+        elasticsearchHost = elasticsearch.host + cloud.host + ":" + elasticsearchPort;
+        elasticsearchScheme = "https";
+
+        final HostAndPort kibana;
+        if (segments.length > 2) {
+            kibana = HostAndPort.parseHostAndPort(segments[2], "Cloud Id, after decoding, the kibana segment is 'undefined', literally. You may need to enable Kibana in the Cloud UI.");
```

Comment:
Although this pulls forward the error message verbatim from the ruby implementation, could we improve it to also tell the user that they need to get a new cloud id once they have performed this action? If they go and enable Kibana, but then continue to use the same cloud id, they will continue to encounter this error.

(cc: @karenzone)